### ISSUE TITLE
perf(float): FP16 conv is slower than FP32 below 4 input channels — the small-channel FP16 kernel regresses on the most common first layers

### ISSUE THREAD

**Issue by apage224** [AmbiqAI/ns-cmsis-nn#480]
## Summary

On Cortex-M55 the FP16 convolution kernel is **slower than FP32 below four input channels**, and the vector path only engages from four channels up. One and three input channels are the audio/grayscale and RGB first layers, so this lands on the entry layer of most TinyML models.

Follow-up to #417, measured on **v7.32.0**. The fully-connected half of #417 is fixed (see the bottom).

## Setup

Apollo510 EVB, Cortex-M55 @ 96 MHz (LP), Arm GNU Toolchain 15.2.Rel1, ns-cmsis-nn **v7.32.0** (`aaeb145a`) built through heliaRT 1.20.0 with `ARM_NN_ENABLE_F32` / `ARM_NN_ENABLE_F16`. Profiled with helia-profiler: single-operator models so nothing is confounded, 100 iterations, median, per-layer PMU counters, arena and weights in TCM unless noted.

Models are one op each: `CONV_2D` 3x3 `padding=same`, 32x32 input, 8 output channels, sweeping input channels; and `FULLY_CONNECTED` 1024 -> N. Built with Keras + `TFLiteConverter` for FP32, then cast to an all-`FLOAT16` graph. Happy to attach the eight `.tflite` fixtures.

## Convolution: the vector path does not engage below C_in = 4

| C_in | MACs | f32 cyc/MAC | f16 cyc/MAC | f16 speedup | total instr f16/f32 | MVE instr f32 -> f16 | MVE stalls f32 -> f16 |
|---|---|---|---|---|---|---|---|
| 1 | 73,728 | 3.45 | 3.73 | **0.93x** | 1.00x | 43,105 -> 69,177 (**1.60x**) | 5,741 -> 37,230 (6.5x) |
| 2 | 147,456 | 2.90 | 3.49 | **0.83x** | 1.16x | 78,073 -> 109,785 (**1.41x**) | 19,275 -> 71,040 (3.7x) |
| 3 | 221,184 | 2.35 | 2.92 | **0.80x** | 1.18x | 106,273 -> 152,649 (**1.44x**) | 20,402 -> 100,352 (4.9x) |
| 4 | 294,912 | 3.48 | 2.63 | 1.32x | 0.68x | 235,789 -> 195,513 (0.83x) | 74,719 -> 129,666 (1.7x) |
| 8 | 589,824 | 2.35 | 1.78 | 1.32x | 0.73x | 419,729 -> 233,357 (0.56x) | 83,546 -> 127,702 (1.5x) |
| 16 | 1,179,648 | 1.75 | 1.25 | 1.40x | 0.65x | 787,609 -> 417,297 (0.53x) | 137,297 -> 228,664 (1.7x) |

Absolute cycles per inference, f32 -> f16: 254,651 -> 274,777 (C_in 1), 427,752 -> 514,624 (2), 519,316 -> 645,720 (3), 1,026,633 -> 776,822 (4), 1,385,852 -> 1,052,542 (8), 2,067,122 -> 1,480,021 (16).

**Three observations.**

1. **Below four channels FP16 issues 40-60 % *more* vector instructions than FP32** and stalls 3.7x to 6.5x as often, which is what makes it slower. At C_in = 1 the total retired instruction count is identical to FP32 (207,131 vs 207,741), so the same scalar work is being done with strictly more vector overhead on top. At C_in = 2 and 3, FP16 retires 16-18 % more instructions in total.
2. **From four channels up the kernel behaves as expected**: FP16 roughly halves the vector instruction count, consistent with 8-lane f16 against 4-lane f32.
3. **FP32 has its own anomaly at C_in = 4**: 3.48 cyc/MAC against 2.35 at both C_in = 3 and C_in = 8, with instruction count jumping from 412,599 to 849,152. Something switches strategy there and the four-channel case is the loser in both precisions.

## Stalls cap the gain even where the path works

At C_in = 16, FP16 halves its vector instruction count but returns only 1.40x, stalling 228,664 times against FP32's 137,297. The pattern holds across every shape where FP16 wins, so the ceiling looks like load-to-use latency or conversion dependencies rather than lane width.

## Fully connected: #417's regression is fixed, and this is what good looks like

| Shape | Weights | f32 cyc/MAC | f16 cyc/MAC | f16 speedup | MVE instr f32 -> f16 |
|---|---|---|---|---|---|
| 1024 -> 12 | TCM | 1.09 | 0.70 | 1.54x | 6,989 -> 3,527 (0.50x) |
| 1024 -> 256 | MRAM | 3.21 | 1.76 | **1.82x** | 148,997 -> 75,202 (0.50x) |

`arm_fully_connected_f16` was ~20 % **slower** than the f32 kernel at v7.31.0 (#417, observation 2). At v7.32.0 it is 1.54x to 1.82x **faster**, with the vector instruction count exactly halved. That is the behaviour the convolution kernel should reach. (The two rows are not comparable to each other: the wide one holds 1 MB of weights in flash. Within each row the comparison is clean.)

## Ask

Make the FP16 convolution kernel take the vectorised path, or at minimum not regress against FP32, for C_in < 4 — and look at the C_in = 4 FP32 cliff while in there. Ambiq's own qualification model is an audio DS-CNN whose first layer is C_in = 1 and holds 58 % of the whole graph's cycles, so this is the single highest-leverage float kernel we have measured.

**Comment by apage224** [comment]
Adding the reproducer, and separating what was measured from what is hypothesis.

## Reproducer (self-contained, no file transfer needed)

Models are one operator each, built with Keras + `TFLiteConverter` for FP32 and cast to an
all-`FLOAT16` graph:

```python
import tensorflow as tf
# conv sweep: C_in in (1, 2, 3, 4, 8, 16)
inp = tf.keras.Input(shape=(32, 32, C_IN), batch_size=1)
out = tf.keras.layers.Conv2D(8, 3, padding="same", use_bias=True)(inp)
open(f"conv_cin{C_IN:02d}_fp32.tflite", "wb").write(
    tf.lite.TFLiteConverter.from_keras_model(tf.keras.Model(inp, out)).convert())

# fully-connected: units in (12, 256)
inp = tf.keras.Input(shape=(1024,), batch_size=1)
out = tf.keras.layers.Dense(UNITS, use_bias=True, bias_initializer="ones")(inp)
```

The FP16 variants are the FP32 files cast to all-`FLOAT16` tensors with
`tools/cast_fp16.py` from AmbiqAI/helia-profiler (not the converter's float16
post-training quantization, which leaves FP32 activations and would not exercise the f16
kernels at all).

Measured with helia-profiler `hpx profile`: `apollo510_evb`, `arm-none-eabi-gcc`
(Arm GNU 15.2.Rel1), RTT, `clock.cpu: lp` (96 MHz), arena 256 KB, weights and arena in TCM
(except FC 1024->256, whose 1 MB of weights sit in MRAM), `iterations: 100`, `warmup: 10`,
`aggregation: median`, per-layer PMU with the CPU and MVE counter groups. heliaRT 1.20.0
(`edb3a25f`) source build against ns-cmsis-nn v7.32.0 (`aaeb145a`) with `ARM_NN_ENABLE_F32`
and `ARM_NN_ENABLE_F16` set before the module is added.

The 18 `.tflite` fixtures and the exact run configs can be attached or shared on request.

## Measured

Everything in the tables of the issue body: cycles, cycles/MAC, total instructions retired,
`MVE_INST_RETIRED`, `MVE_STALL`, per shape and per precision. Specifically:

- FP16 is slower than FP32 at C_in 1, 2, 3 (0.93x / 0.83x / 0.80x) and faster at 4, 8, 16
  (1.32x / 1.32x / 1.40x).
- Below 4 channels FP16 retires 40-60 % more MVE instructions than FP32; at and above 4 it
  retires 17-47 % fewer.
- At C_in = 1 the total retired instruction count is within 0.3 % between precisions
  (207,131 f16 vs 207,741 f32). At C_in = 2 and 3, FP16 retires 16-18 % MORE in total.
- FP32 at C_in = 4 costs 3.48 cyc/MAC against 2.35 at both C_in = 3 and C_in = 8, with
  total instructions jumping 412,599 -> 849,152.
- FULLY_CONNECTED halves its MVE instruction count in FP16 at both widths tested (0.50x).

## Hypothesis, not measured

- That the sub-4-channel path is an im2col or channel-blocking fallback rather than the
  vectorised path. Not instrumented; inferred from the instruction mix.
- That the stall counts reflect load-to-use latency or f16/f32 conversion dependencies. The
  stall counter does not attribute a cause, so this is a guess at the mechanism.
- That the C_in = 4 FP32 cost reflects a strategy switch. Only the cost and the instruction
  jump are measured; nothing identifies which path was taken.

Happy to instrument any of these further on request — the bench is attached and the runs
take about a minute each.

**Comment by apage224** [comment]
## Correction: the vector path does engage; the small-channel kernel is the one that regresses

I framed this as "the vector path does not engage below 4 input channels". That was a
hypothesis from the instruction mix and **it is wrong**. Reading the v7.32.0 sources actually
built for those runs:

- `arm_convolve_wrapper_f16` passes straight through to `arm_convolve_f16(..., NHWC)`. With a
  3x3 filter neither the 1x1 nor the 1xN path applies, so one predicate decides everything.
- `arm_conv_nhwc_use_small_c_f32` selects a dedicated small-channel MVE kernel while
  `input_c < ARM_NN_CONV_SMALL_C_F32_LANES` = **4** (`OC_GROUP` 4).
- `arm_conv_nhwc_use_small_c_f16` does the same while `input_c < ARM_NN_CONV_SMALL_C_F16_LANES`
  = **8** (`OC_GROUP` 2).
- Both guards additionally require the u16 gather reach to fit; for these shapes reach is 42
  (f16) and 38 (f32), so `reach * input_c` peaks at 168 against a 65535 limit. The offset
  guard is not the limiter, the threshold is.

So at C_in 1 and 3 both precisions run their small-C MVE kernel, and the FP16 deficit
(0.93x, 0.80x) is that kernel being slower than the FP32 one, not a scalar fallback.

**The C_in=4 row is also not like-for-like** and I should not have presented it beside the
others: FP32 has crossed to the general NHWC path at 4, while FP16 stays on small-C until 8.
The 1.32x there is mostly FP32 getting worse at its own crossover, 2.35 -> 3.48 cyc/MAC from
C_in 3 to 4 on strictly more work, while FP16 moves only 2.92 -> 2.63.

That FP32 crossover looks mistuned in its own right, and is a separate question from the FP16
comparison.

All measured numbers in the issue body stand unchanged; only the causal reading above is
corrected. One remaining hypothesis, flagged as such: the small-C kernels differ in shape,
FP32 covering 4 lanes x 4 output channels and FP16 8 lanes x 2, so with only 8 output
channels in these shapes FP16 needs more passes over the input. Not measured.

**Comment by apage224** [comment]
## Two corrections from an independent review of this evidence

### The gather-reach guard is FP16-only

My previous comment said "both guards additionally require the u16 gather reach to fit".
That is wrong for FP32. At `aaeb145a`:

```c
__STATIC_INLINE bool arm_conv_nhwc_use_small_c_f32(const cmsis_nn_dims *input_dims,
                                                   const cmsis_nn_dims *output_dims)
```

It takes no `conv_params`, computes no reach, and returns true on the channel and
width checks alone — the FP32 kernel gathers with u32 offsets, so it has nothing to
guard. Only `arm_conv_nhwc_use_small_c_f16` carries the reach test against
`ARM_NN_MVE_F16_GATHER_OFFSET_MAX`. A shape with width 30000 and C_in 3 is therefore
rejected by FP16 and accepted by FP32.

The thresholds themselves are unchanged: FP32 small-C while `input_c < 4`, FP16 while
`input_c < 8`, and for the shapes measured here the offset guard is not the limiter.

### Two preserved FP32 ELFs were the wrong build

The artifacts I kept for the C_in 3 and C_in 4 FP32 cases held a different build:
`conv_cin03_fp32.elf` was byte-identical to `conv_cin01_fp16.elf`, and
`conv_cin04_fp32.elf` to `conv_cin03_fp16.elf`. My capture step picked the build to
copy by modification time, and FP32 and FP16 configurations occupy separate
dependency workspaces, so once an FP16 build existed the FP32 copies took it.

Both are repaired: C_in 4 recovered from its workspace, C_in 3 regenerated. Each is
now verified by extracting the bytes of `_ZL10model_data` from the ELF and comparing
them with the complete fixture, and the capture step now selects the build from the
run's own recorded workspace fingerprint and refuses an artifact that fails that
check.

**The measurements in this issue are unaffected.** Each run records
`firmware.model_size` as reported over the wire by the image that was executing, and
all six agree with their intended fixture (1388, 1244, 1964, 1532, 2260, 1684 bytes).
Re-running C_in 3 FP32 three hours later gives 519,315 cycles against the 519,316 in
the table, with the same 518,784 clean end-to-end and 221,184 MACs.

Thanks to the reviewer who caught this by extracting the embedded models rather than
trusting my filenames.

**Comment by apage224** [comment]
# CORE #480: preserve accuracy while improving small-channel FP16 convolution

Adam selected this work on2026-09-10 while TESTER completes133 and the paired496 descriptors. Base: published7.33.1 adeb1f90c70d6c601237f7dcda5ccf8dbac702f9. Author worktree: workspaces/core/worktrees/ns-cmsis-nn-480-small-channel, branchperf/480-small-channel-f16. GATHER496 stays frozen/draft and is not part of this change.

## Demonstrated starting point

The current f16 convolution source/sizer and optimization header are byte-identical to measured7.32 baselineaaeb145a. Small-C already uses MVE and FP32 accumulation. Cin1/Cin3 current baseline executes against the preserved fixture weights and real LiteRT reference with half-rounded operands, at input ranges[-1,1] and[-64,64]; all outputs match reference half bits in host sanitizers and M55GCC14.2 O2/Ofast. Existing HPX geometry is32x32,Cout8,3x3SAME, stride/dilation1 with bias. Previous threshold bypass exposed half accumulation and was rejected; it is not a candidate.

## Bounded candidate and hypothesis

Change the internal FP16 small-C compute tile from eight spatial positions × two output channels to four positions × four output channels. This keeps four FP32 accumulators and reduces the fraction of positions requiring edge handling for common widths. Use MVE halfword loads/gathers extended into32-bit lanes, then the existing half-to-float widening helper; apply one32-bit predicate to loads and FMAs. Narrow once per output, preserve half activation clamp order. This removes even/odd predicate splitting and may lower edge overhead; an improvement is a hypothesis until matched measurements.

Retain the existing small-C admission predicate (Cin<8 and original reach guard), public signatures, reported scratch sizes, layouts, supported strides/dilations and all fallback arithmetic. Spatial tile size must be separate from input-channel admission and NT_N_PACKED's fixed eight-channel weight-block width. Four-channel groups must not cross packed blocks; partial groups load only live channels. For Cin1 stride1, retain bounded contiguous loads where valid; otherwise use bounded gathers. No NEON, scalar half ternaries, FP16 accumulation or tolerance changes. Do not change FP32 thresholds as part of this candidate; FP32 is a measured control for the reported cliff.

The new u32 halfword MVE gather/scatter intrinsics compile with GCC14.2 onM55/F16-only; see INTRINSIC-PROBE.json. Widened half loads follow the existing dequantize_f16_f32 idiom. Code may be simplified after measured results, but do not add a second public dispatch API or broad framework.

## Validation and acceptance

1. Independent plan review before kernel edit. Reuse preserved real oracle data, not baseline outputs as correctness reference.
2. Candidate must preserve current output bits on same-build small-C regression inputs and pass the independent reference, including finite cancellation/range controls. Exercise Cin1,2,3,4,7; output channel tails and packed block crossings; widths around4/8 and edges; batches, bias/no bias, stride/dilation and SAME/VALID. Exhaust all65536half input encodings in bounded known-outcome convolution cases, and exercise NaN/Inf/padded-tap behavior and FP controls without changing tolerance.
3. Host sanitizers; M55O2/O3/Ofast and autovectorization; dtype gates; GCC13.2 floor; applicable M4/M0 FP32 regression compile/execution and QEMU. Record exact effective flags and runtime limitations. Mutate the changed computation or predicates to prove regression tests fail.
4. Inspect actual compiled route/register spills and stable source/ELF/model identity. After correctness passes, ask HPX for matched baseline/candidate Cin1 andCin3 captures on Apollo510, three runs/median, same clock/cache/placement/compiler and original fixtures; include Cin2/Cin4 boundary controls if bounded. No board claim or throughput ratio from simulator wall time. Retain a candidate only for demonstrated performance benefit without material regression on the agreed cases.
5. Two complementary independent adversarial reviews, fixes/re-review of affected scope, judgment and final light review. Keep draft while iterating. Reuse existing tester cases where they actually hit the changed path; prepare direct CORE tests locally. Do not interrupt TESTER133/496 or add descriptor requirements without a concrete measured coverage gap. One fullCI after local evidence, separate Adam merge approval.

## Limits and stop conditions

The half-accumulation precision limitation in general/packed matmul is documented separately and remains outside this performance patch. A performance candidate cannot claim to fix it. No new operator family, release-tooling changes or consumer changes. Stop/rework if oracle accuracy, guards, target gates or matched performance regress; no blind threshold tuning or tolerance relaxation.
